### PR TITLE
bpo-45138: Expand traced SQL statements in `sqlite3` trace callback

### PR DESCRIPTION
**Pull request by erlend-aasland**
https://bugs.python.org/issue45138

### CODE REVIEW

File: Modules/_sqlite/connection.c [hunk 1]
```diff
@@ -1074,11 +1073,37 @@ trace_callback(void *ctx, const char *statement_string)
...
+            goto exit;
+        }
+
+        pysqlite_state *state = ((callback_context *)ctx)->state;
+        assert(state != NULL);
+        if (state->enable_callback_tracebacks) {
+            PyErr_SetString(state->DataError,
+                            "Expanded SQL string exceeds the maximum string "
+                            "length");
+            PyErr_Print();
```
Comment (pablogsal) [inline Modules/_sqlite/connection.c]: Why this? If this is an unraisable exception you may want to use `PyErr_WriteUnraisable`.

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: Ah, I was unaware of that C API function. Nice; thanks!

We've never raised exceptions in the trace handler. I had a discussion with Petr regarding this on https://github.com/python/cpython/pull/28133#issuecomment-914286495.

Also, `sqlite3` has this "print an exception or clear it" option before signalling UDF errors back to SQLite.

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: See b91fc81849479a4368e21cdff413e7949cd944a4. Maybe we can modify `print_or_clear_traceback()` to use `PyErr_WriteUnraisable` later.

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: I rewrote this in b91fc81849479a4368e21cdff413e7949cd944a4. PTAL.

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: FYI, I've created a WIP branch that modifies`print_or_clear_traceback()` to use unraisable exceptions. I prefer to land this PR using the old/current behaviour before possibly changing the callback code to use unraisable exceptions.

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: I'm marking this conversation as resolved. Feel free to undo if you disagree :)

**Comment by erlend-aasland** [comment]
I'll resolve the conflicting files after GH-29327 has been merged.

**Comment by erlend-aasland** [comment]
Note to self: after GH-29356 has landed, add test for too long SQL statements.

**Comment by erlend-aasland** [comment]
Coverage for `trace_callback()` is now pretty good.

<details>

```
 1056|       |static int
 1057|       |trace_callback(unsigned int type, void *ctx, void *stmt, void *sql)
 1058|       |#else
 1059|       |static void
 1060|       |trace_callback(void *ctx, const char *sql)
 1061|       |#endif
 1062|     20|{
 1063|     20|#ifdef HAVE_TRACE_V2
 1064|     20|    if (type != SQLITE_TRACE_STMT) {
 1065|      0|        return 0;
 1066|      0|    }
 1067|     20|#endif
 1068|       |
 1069|     20|    PyGILState_STATE gilstate = PyGILState_Ensure();
 1070|       |
 1071|     20|    assert(ctx != NULL);
 1072|     20|    PyObject *py_statement = NULL;
 1073|     20|#ifdef HAVE_TRACE_V2
 1074|     20|    assert(stmt != NULL);
 1075|     20|    const char *expanded_sql = sqlite3_expanded_sql((sqlite3_stmt *)stmt);
 1076|     20|    if (expanded_sql == NULL) {
 1077|      2|        sqlite3 *db = sqlite3_db_handle((sqlite3_stmt *)stmt);
 1078|      2|        if (sqlite3_errcode(db) == SQLITE_NOMEM) {
 1079|      0|            (void)PyErr_NoMemory();
 1080|      0|            goto exit;
 1081|      0|        }
 1082|       |
 1083|      2|        pysqlite_state *state = ((callback_context *)ctx)->state;
 1084|      2|        assert(state != NULL);
 1085|      2|        PyErr_SetString(state->DataError,
 1086|      2|                        "Expanded SQL string exceeds the maximum string "
 1087|      2|                        "length");
 1088|       |
 1089|       |        // Fall back to unexpanded sql
 1090|      2|        py_statement = PyUnicode_FromString((const char *)sql);
 1091|      2|    }
 1092|     18|    else {
 1093|     18|        py_statement = PyUnicode_FromString(expanded_sql);
 1094|     18|        sqlite3_free((void *)expanded_sql);
 1095|     18|    }
 1096|       |#else
 1097|       |    py_statement = PyUnicode_FromString(sql);
 1098|       |#endif
 1099|     20|    if (py_statement) {
 1100|     20|        PyObject *exc, *val, *tb;
 1101|     20|        PyErr_Fetch(&exc, &val, &tb);
 1102|       |
 1103|     20|        PyObject *callable = ((callback_context *)ctx)->callable;
 1104|     20|        PyObject *ret = PyObject_CallOneArg(callable, py_statement);
 1105|     20|        Py_DECREF(py_statement);
 1106|     20|        Py_XDECREF(ret);
 1107|       |
 1108|     20|        _PyErr_ChainExceptions(exc, val, tb);
 1109|     20|    }
 1110|       |
 1111|     20|    if (PyErr_Occurred()) {
 1112|      4|        print_or_clear_traceback((callback_context *)ctx);
 1113|      4|    }
 1114|       |
 1115|     20|exit:
 1116|     20|    PyGILState_Release(gilstate);
 1117|     20|#ifdef HAVE_TRACE_V2
 1118|     20|    return 0;
 1119|     20|#endif
 1120|     20|}
```

</details>

File: Modules/_sqlite/connection.c (hunk 1 above)
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: I opened [bpo-45828](https://bugs.python.org/issue45828) and GH-29591 for this.

**Comment by erlend-aasland** [comment]
@pablogsal PTAL

File: Modules/_sqlite/connection.c [hunk 2]
```diff
@@ -1112,24 +1111,51 @@ trace_callback(void *ctx, const char *statement_string)
...
+        }
+
+        pysqlite_state *state = ((callback_context *)ctx)->state;
+        assert(state != NULL);
+        PyErr_SetString(state->DataError,
+                        "Expanded SQL string exceeds the maximum string "
+                        "length");
+
+        // Fall back to unexpanded sql
+        py_statement = PyUnicode_FromString((const char *)sql);
```
Comment (pablogsal) [inline Modules/_sqlite/connection.c:1114]: Hummm, this is technically wrong: you cannot call the C-API with an error set. Some of the API-s may even hard-fail:

https://github.com/python/cpython/blob/4141d94fa608cdf5c8cd3e62f7ea1c27fd41eb8d/Objects/unicodeobject.c#L11507

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:1114]: Ah, yes of course. I'll update my PR.

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:1114]: PTAL with 391c698b2239a4b32a9b23c3ed0ee1b0f95ae561 applied.

**Comment by erlend-aasland** [comment]
I believe I've addressed everything here, @pablogsal. Let me know if you want further changes.

File: Lib/test/test_sqlite3/test_hooks.py [hunk 3]
```diff
@@ -20,12 +20,27 @@
...
+import unittest
 
 from test.support.os_helper import TESTFN, unlink
+
+from test.test_sqlite3.test_dbapi import memory_database, cx_limit
 from test.test_sqlite3.test_userfunctions import with_tracebacks
 
+
+@contextlib.contextmanager
+def check_stmt_trace(self, cx, expected):
```
Comment (JelleZijlstra) [inline Lib/test/test_sqlite3/test_hooks.py]: Makes more sense to make this a method on the test class.

File: Lib/test/test_sqlite3/test_hooks.py [hunk 4]
```diff
@@ -20,12 +20,27 @@
...
+    try:
+        traced = []
+        cx.set_trace_callback(lambda stmt: traced.append(stmt))
```
Comment (JelleZijlstra) [inline Lib/test/test_sqlite3/test_hooks.py]:
```suggestion
        cx.set_trace_callback(traced.append)
```

File: Lib/test/test_sqlite3/test_hooks.py [hunk 5]
```diff
@@ -289,6 +304,51 @@ def trace(statement):
...
+    @with_tracebacks(
+        sqlite.DataError,
+        regex="Expanded SQL string exceeds the maximum string length"
+    )
+    def test_trace_too_much_expanded_sql(self):
+        # If the expanded string is too large, we'll fall back to the
+        # unexpanded SQL statement. The resulting string length is limited by
+        # SQLITE_LIMIT_LENGTH.
+        template = "select 'b' as \"a\" from sqlite_master where \"a\"="
+        category=sqlite.SQLITE_LIMIT_LENGTH
```
Comment (JelleZijlstra) [inline Lib/test/test_sqlite3/test_hooks.py]:
```suggestion
        category = sqlite.SQLITE_LIMIT_LENGTH
```

File: Modules/_sqlite/connection.c [hunk 6]
```diff
@@ -1092,24 +1091,46 @@ trace_callback(void *ctx, const char *statement_string)
...
+    }
+    else {
+        py_statement = PyUnicode_FromString(expanded_sql);
+        sqlite3_free((void *)expanded_sql);
+    }
+#else
+    py_statement = PyUnicode_FromString(sql);
```
Comment (JelleZijlstra) [inline Modules/_sqlite/connection.c:1121]: Looks like this will use "strict" as the errors handler, while the old code had "replace". Does that matter?

**Review by JelleZijlstra — COMMENTED** [review]
Thanks, a few small comments

File: Lib/test/test_sqlite3/test_hooks.py (hunk 4 above)
Comment (erlend-aasland) [reply Lib/test/test_sqlite3/test_hooks.py]: Doh 🤦🏻

File: Lib/test/test_sqlite3/test_hooks.py (hunk 3 above)
Comment (erlend-aasland) [reply Lib/test/test_sqlite3/test_hooks.py]: I agree

File: Modules/_sqlite/connection.c (hunk 6 above)
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:1121]: It does not matter, AFAICS; the trace callback is only invoked on successfully prepared statements, which implies that the statement is already valid UTF-8. I don't see how the expanded part could end up as invalid UTF-8 (unless you've got a broken SQLite library installed).

**Review by JelleZijlstra — APPROVED** [review]
Looks good! @gvanrossum I'm planning to merge this change, another SQLite enhancement by Erlend.

**Comment by erlend-aasland** [comment]
Thanks, @JelleZijlstra!

**Comment by kumaraditya303** [comment]
This change broke several buildbots See https://buildbot.python.org/all/#/builders/58/builds/1850

**Comment by erlend-aasland** [comment]
I'm aware, and I'm coming up with a fix for that test.

**Pull request merged**